src/qibolab/platforms/multiqubit.py

```diff
@@ -177,9 +186,108 @@ def execute_pulse_sequence(self, sequence: PulseSequence, nshots=None):
 
         for ro_pulse in ro_pulses.values():
             data[ro_pulse.serial] = ExecutionResults.from_components(*acquisition_results[key])
-            data[ro_pulse.qubit] = copy(data[ro_pulse.serial])
+            data[ro_pulse.qubit] = copy.copy(data[ro_pulse.serial])
         return data
 
+    def sweep(self, sequence, *sweepers, nshots=1024, average=True, wait_time=None):
+        results = {}
+        map_sweepers = {}
+        copy_sequence = copy.deepcopy(sequence)
+        for sweeper in sweepers:
+            map_sweepers[sweeper.parameter] = {
+                pulse.serial: pulse for pulse in copy_sequence if pulse in sweeper.pulses
+            }
+
+        self._sweep_recursion(
+            copy_sequence,
+            copy.deepcopy(sequence),
+            *sweepers,
+            nshots=nshots,
+            average=average,
+            wait_time=wait_time,
+            results=results,
+            map_sweepers=map_sweepers,
+        )
+        return results
+
+    def _sweep_recursion(
+        self,
+        sequence,
+        original_sequence,
+        *sweepers,
+        nshots=1024,
+        average=True,
+        wait_time=None,
+        results=None,
+        map_sweepers=None
+    ):
```

Comment:
This function is definitely too complex: you should consider to split the various sections and give them a meaningful name (and possibly docstring).

Reply:
I know. I'm not very proud of this implementation. I think that there is surely a better way of implementing it. For time being I will try to break the function in several parts.

Reply:
> For time being I will try to break the function in several parts.

I know it seems dumb, but already breaking helps in non-trivial names:
- giving names to blocks is self-documenting, so it is doing the job of the comments headers, even when you are not actually reusing the functions
- declaring explicitly the function parameters is telling you which parameters are used in which block, while if you have a single big functions they are always all in scope

Reply:
Fixed.